src/util/task-queue.js

```diff
@@ -46,10 +48,37 @@ class TaskQueue {
      *
      * @param {Function} task - the task to run.
      * @param {number} [cost=1] - the number of tokens this task consumes from the bucket.
-     * @returns {Promise} - a promise for the task's return value.
+     * @returns {Promise} - a promise for the task's return value, or `undefined` if the task gets canceled.
      * @memberof TaskQueue
      */
     do (task, cost = 1) {
+        return this._do(task, false, cost);
+    }
+
+    /**
+     * Wait until the token bucket is full enough, then run the provided task.
+     *
+     * @param {Function} task - the task to run.
+     * @param {number} [cost=1] - the number of tokens this task consumes from the bucket.
+     * @returns {Promise} - a promise for the task's return value. This promise will reject if the task gets canceled.
+     * @memberof TaskQueue
+     */
+    doOrReject (task, cost = 1) {
+        return this._do(task, true, cost);
+    }
+
+    /**
+     * Wait until the token bucket is full enough, then run the provided task.
+     *
+     * @param {Function} task - the task to run.
+     * @param {boolean} rejectOnCancel - choose what happens if the queue cancels this task:
+     *   true: reject the promise returned by this function
+     *   false: resolve the promise returned by this function with a value of `undefined`
+     * @param {number} [cost=1] - the number of tokens this task consumes from the bucket.
+     * @returns {Promise} - a promise for the task's return value.
+     * @memberof TaskQueue
+     */
+    _do (task, rejectOnCancel, cost) {
```

Comment:
Would it look okay for this to be called from outside the file?  I think the hardware extensions want to call this function more often than `doOrReject`

Reply:
I was thinking you'd want to call either of these:
```js
taskQueue.do(() => { stuff(); }, someCost);
taskQueue.doOrReject(() => { stuff(); }, someCost).catch(...);
```
...both of which are exposed currently. If you'd like to call this `_do` method that's fine, but I expect you'd rather call [`do()` above](https://github.com/LLK/scratch-vm/pull/2033/files#diff-ea8d71530e1deef7ef896b9e29914cfbR54).

Reply:
Oh sorry!  I didn't see the full context and realize `do()` was still available, I was not paying attention and only going by the changes here.  Thank you for clarifying!  Sounds good.